### DISCUSSION TITLE
weird top alignment only in the TOC of Bookdown's gitbook output

### DISCUSSION THREAD

**Discussion by thosgood** [KaTeX/KaTeX#3110] [Q&A]
> Before reporting a bug
- [x] Check [common issues](https://katex.org/docs/issues.html).
- [ ] ~~Check the bug is reproducible in [the demo](https://katex.org). If not, check KaTeX is up-to-date and installed correctly.~~
- [x] Search for [existing issues](https://github.com/KaTeX/KaTeX/issues).

**Describe the bug:**
I am using Bookdown's gitbook output to generate html pages. Normally, these use MathJax to render maths, but I've switched over to KaTeX, and everything works beautifully within the pages themselves, but there is something odd about any KaTeX rendering in the table of contents, namely certain elements (such as `\widehat{\otimes}`, or even just `\mathrm{C}_0(M)`) are invisible. Digging around in the browser inspector, I can get them to reappear if I disable the `top` CSS attribute (e.g. `top: -2.55em` for the `0` in `\mathrm{C}_0(M)`). I cannot, however, find a way to fix all the missing elements in one go, and I'm not sure if there is some clever CSS I can use, or if I need to render things differently (my render JS can be found below). Any help is much appreciated!

**Screenshots:**
You can view the live version of the screenshot below at [https://labs.thosgood.com/translations/AIF-4-1952-73.html](https://labs.thosgood.com/translations/AIF-4-1952-73.html).

<img width="291" alt="Screenshot 2021-07-07 at 16 12 15" src="https://user-images.githubusercontent.com/7130759/124784866-409e1380-df3e-11eb-8052-2c139d3f5c37.png">

**(La)TeX code:**
The code of (La)TeX you tried to render:
```latex
E\widehat{\otimes} F
```

**Rendering code:**
```javascript
var $document = $(document);
$document.ready(function() {
  $('eq').map(function() {
    var display = (this.getAttribute('env') == "displaymath");
    katex.render(this.textContent, this, {displayMode: display, fleqn: true, throwOnError: false});
  });
});
```

**Environment (please complete the following information):**
 - KaTeX Version: 0.13.11
 - Device: Desktop
 - OS: OS X
 - Browser: Firefox
 - Version: 89.0.2

**Comment by edemaine** [comment]
Thanks for including a public version of the website! It has a lot of CSS rules of the form `.book ... span` which are impacting all KaTeX elements.  This won't work, as KaTeX uses CSS a lot to fix layout.  Is it possible to remove those rules, or make them more specific (e.g. adding `>` before `span` so that it's just direct children)?

**Reply by thosgood** [reply]
unfortunately, I don't have any control over these rules, since they are automatically generated via bookdown/gitbook. I suppose it might be possible to rewrite them dynamically with javascript, but I'm not sure...

**Reply by ronkok** [reply]
KaTeK spans are all descendants of the `.katex` class, so another possible approach would be to write a CSS reset that insulates every KaTeX span from outside influences. Something like that is already in place for the nested SVGs, but not for generic spans.

A sizable effort, I would think.

**Reply by thosgood** [reply]
hm, I'm not sure how exactly I would go about doing this...
(as a temporary "fix", I've simply opted for showing the `katex-mathml` part instead of the `katex-html`; there are still some rendering problems, but much less severe)

<img width="287" alt="Screenshot 2021-07-10 at 04 00 27" src="https://user-images.githubusercontent.com/7130759/125149844-64767a80-e133-11eb-8239-5c7a1847aa17.png">

**Reply by edemaine** [reply]
MathML doesn't work on Chrome yet, so beware that the math won't render there.

Something you could try is adding `!important` to all the KaTeX CSS rules. That should be enough to override some of bookdown/gitbook's rules (I recall `display` rules in particular, which this would help fix). Then for properties set by bookdown/gitbook but not KaTeX, @ronkok's suggesting a rule like `.katex { margin: default !important; padding: default !important; }`.

**Reply by thosgood** [reply]
unfortunately this doesn't work either (nor any one of the many things that I tried --- the closest I could get was manually changing every span to `display: inline-block`, and that _sort of_ worked, but for some reason wouldn't work if I just used `!important`, and setting `!important` in all of `katex.css` just messed things up in the body of the page)

For now I've disabled KaTeX rendering just in the TOC, and, worst case, I might have to look at using MathJax or something to do the TOC separately (which is obviously a bit annoying).

I'm guessing there's no way for me to tell KaTeX to render in such a way that it puts all the CSS styling stuff inline? As in, a way of calling KaTeX without having to include `katex.css`? This would solve the problem I think (since inline styling always takes priority), but I doubt it's common functionality!

**Reply by edemaine** [reply]
Sadly, no, that's not a feature. I've wondered about it for rendering HTML email (though I'm not sure the fonts would be respected in email, and fonts require CSS). Plausibly you could [do some postprocessing](https://stackoverflow.com/questions/15667426/is-it-possible-to-check-if-certain-css-properties-are-defined-inside-the-style-t/15669079#15669079) to see which KaTeX rules apply and then inline-style them...

**Reply by edemaine** [reply]
One other thing: perhaps you could raise an issue with Bookdown, to encourage a rewrite of that CSS rule to be more specific?
